query-engine/core/src/query_graph_builder/write/update.rs

```diff
@@ -131,30 +168,76 @@ pub fn update_many_records(
     Ok(())
 }
 
-/// Creates an update record query node and adds it to the query graph.
+/// Creates an update record write node and adds it to the query graph.
 pub fn update_record_node<T: Clone>(
     graph: &mut QueryGraph,
     query_schema: &QuerySchema,
     filter: T,
     model: Model,
     data_map: ParsedInputMap<'_>,
+    field: Option<&ParsedField<'_>>,
 ) -> QueryGraphBuilderResult<NodeRef>
 where
     T: Into<Filter>,
 {
-    graph.flag_transactional();
-
     let update_args = WriteArgsParser::from(&model, data_map)?;
     let mut args = update_args.args;
 
     args.update_datetimes(&model);
 
     let filter: Filter = filter.into();
-    let update_parent = Query::Write(WriteQuery::UpdateRecord(UpdateRecord {
-        model: model.clone(),
-        record_filter: filter.into(),
-        args,
-    }));
+
+    // If the connector can use `RETURNING`, always use it as it may:
+    // 1. Save a SELECT statement
+    // 2. Avoid from computing the ids in memory if they are updated. See `update_one_without_selection` function.
+    let update_parent = if query_schema.has_capability(ConnectorCapability::UpdateReturning) {
+        let has_nested_selection = field.map(|f| f.has_nested_selection()).unwrap_or(false);
+
+        // If there are nested operations or nested selections, then make the graph transactional.
+        // Otherwise, it means the operation can be done in a single, atomic update.
+        if !update_args.nested.is_empty() || has_nested_selection {
```

Comment:
I love these comments, the code is really clear and well written. But the inline commments provide additional clarity. Thank you.